### ISSUE TITLE
`terminate` not called on `Ctrl+C`

### ISSUE THREAD

**Issue by relf** [argmin-rs/argmin#305]
I've successfully argminified my solver [here](https://github.com/relf/egobox/pull/67) (thanks again for your work). 

I've used the ability to interrupt the executor using `Ctrl+C` which indeed ends the current iteration but the `terminate` method does not seem to be called at the end. As `terminate` is documented as being called after each iteration somehow I was expecting it to be also called in this case to end gracefully.  

Looking at the executor code, I was wondering if a terminaison reason `KeyboardInterrupt` could be added to handle this specific case seperately from `Aborted`. 

Maybe a better option would be to add a `final_terminate` method called only once at the end of `run` (as a symetric of `init`) to allow whatever cleanup is needed. What do you think?

**Comment by stefan-k** [comment]
The purpose of `terminate` is to assess whether or not the algorithm should be stopped from iterating and feed that back to the `Executor` loop. `Ctrl+C` is a bit different however, because it directly stops the loop from progressing to the next iteration. Therefore, in case of `Ctrl+C`, the last call to the terminate function will be the one in the last loop iteration that was still executed [1].

I like the idea of a final method (which would be part of the `Solver` trait) that has mutable access to the state and the problem and which is called right before returning from `run`. But I would probably call it `finalize` instead of `final_terminate`, because it wouldn't return a `TerminationReason`. Would that help you in what you are trying to achieve?  

> Looking at the executor code, I was wondering if a terminaison reason KeyboardInterrupt could be added to handle this specific case seperately from Aborted.

Those two would then be equivalent, because `Aborted` is only used for `Ctrl+C` (so far). 

[1] I wonder if the `Ctrl+C` handling could be moved into `terminate_internal`.

**Comment by relf** [comment]
Yes `finalize` is the right term.

> Would that help you in what you are trying to achieve?

At the moment my point was to only log the interruption, just to know that it was intentionally interrupted. With the `finalize` maybe I will be able to make the same check as you when you set the termination reason to `Aborted` to log the keyboard interruption. To my understanding only `Ctrl+C` is able to set the atomic bool `running` to `false`, so the `Executor` could properly handle the termination reason KeyboardInterrupt by testing this boolean after the loop. Did I miss something?

> Those two would then be equivalent, because Aborted is only used for Ctrl+C (so far).

Actually, I came up to use `Aborted` when my solver is unable to find a suitable `param` after several retries, and I discovered afterwards that I was unable to distinguish it from keyboard interruption (hence my suggestion). But I can also workaround this by choosing another termination like `AcceptedStallIterExceeded` as it is a kind of stall (but the note "for Simulated Annealing" stopped me in the first place).

Thinking about `TerminationReason`, I would suggest to distinguish the status `NotTerminated/Terminated` and only in the second case have the reason. Something like:

```rust
enum TerminationStatus {
   NotTerminated,
   Terminated(TerminationReason)
}
```
It would avoid the "weird" `TerminationReason::NotTerminated`. And maybe the `terminate_internal` could be renamed `(get_)termination_status` and the recommended user-facing `terminate` something like `assess_termination_status` as you described it. 

> I wonder if the Ctrl+C handling could be moved into terminate_internal

The handler has to be set once so I am not sure what you mean when you want to move it in `terminate_internal`.

**Comment by stefan-k** [comment]
Thanks for this extended explanation and your excellent ideas. I particularly like the `TerminationStatus` idea because I never was a fan of the obviously weird `TerminationReason::NotTerminated`. I wonder if turning `TerminationReason` into a trait and having the following would be even better:

```rust
enum TerminationStatus<T> {
   NotTerminated,
   Terminated(T)
}
```
where `T: TerminationReason`.
I don't have all the code in my head and therefore I don't know if this potentially causes problems with generics, so it may be necessary to solve it like this:

```rust
enum TerminationStatus {
   NotTerminated,
   Terminated(Box<dyn TerminationReason>)
}
```

That way, termination reasons can be created for each solver independently. So far only `argmin` can define termination reasons which is an unpleasant limitation. An obvious downside would be that one cannot match on termination reasons anymore. 

I now agree that there should be a `KeyboardInterrupt` reason as well.

> To my understanding only Ctrl+C is able to set the atomic bool running to false, so the Executor could properly handle the termination reason KeyboardInterrupt by testing this boolean after the loop. Did I miss something?

This is already done, but in a more awkward way (right before `run` returns). Placing a `finalize` function call right after this should work for you, because the `TerminationReason::KeyboardInterrupt` will be set in the state and hence would be available in `finalize`. 

> The handler has to be set once so I am not sure what you mean when you want to move it in terminate_internal.

The handler would still be set in `run` but the check whether `running` is false would be done in `terminate_internal`. But thinking about it in more detail I realized that this isn't a great idea.

**Comment by relf** [comment]
Creating termination reason per solver seems appealing but I am not sure about the added complexity of using generic or trait objects. I mean in the first place I was just happy with `Aborted`, as I have no use case on my side where I would like to handle the termination reason to do anything else than printing/logging the reason (which can be done by the solver itself anyway). I am wondering if just broad termination categories like now (may be with a string for details) wouldn't be just enough... Just my feeling on this.

**Comment by stefan-k** [comment]
Generics probably won't work. I think the added complexity of trait objects is probably manageable but you're right that it is likely not really worth the effort. As an alternative, I could add another `Other(String)` to the `TerminationReason` enum to cover special cases. 

I plan to extend the observer interface in the future to make it more flexible. One possible addition could be a `on_ctrlc` method which could be used for exactly that. 

For now I think the following needs to be done:

* Add a `finalize` method to the `Solver` trait and call this method right at the end of `Executor`s `run`.
* Add `TerminationReason::KeyboardInterrupt` and change current uses of `Aborted` to `KeyboardInterrupt`
* Add `TerminationReason::Other(String)`
* Add `TerminationStatus` 

The first two should solve your issue. 

I hope I'll be able to work on this soon, but you're of course also welcome to have a go at this (or parts of it) if you want to!

Thanks for your feedback, I really appreciate it! :)

**Comment by relf** [comment]
Thanks! And ok I can work on these points.

**Comment by stefan-k** [comment]
Sorry I wasn't able to reply to this earlier.

Refactoring `TerminationReason` sounds good to me. 

> I am not familiar with the various solvers but the difference between `TargetPrecisionReached` and `TargetToleranceReached` here is not that meaningful to me.

Thats true, there isn't really a clear distinction between the two, so we might as well remove one.
`TargetToleranceReached` is used in NewtonCG, GoldenSectionSearch, and NelderMead. I think in all three it can be replaced with `TargetPrecisionReached`.
The same applies to `NoChangeInCost`. This is, as far as I can see, also equivalent to `TargetPrecisionReached`. 

Maybe we should think about renaming `TargetPrecisionReached` to `Converged`. If we do rename it, I would even replace `LineSearchConditionMet` with `Converged`. 

My suggestion would be a small extension of your first suggestion:

```rust
pub enum TerminationReason {
    /// Reached maximum number of iterations
    MaxItersReached,
    /// Reached target cost function value
    TargetCostReached,
    /// Converged
    Converged
    /// Algorithm manually interrupted with Ctrl+C
    KeyboardInterrupt,
    /// Solver exit as given condition is met
    SolverExit(String)
}
```

One could also argue that `TargetCostReached` is also somehow `Converged`; however I think it makes sense to be able to distinguish between the two. 

What do you think?

Luckily `TerminationReason` isn't `#[non_exhaustive]` so this breaking change should be noticed by all downstream crates.

**Comment by relf** [comment]
`MaxItersReached`, `TargetCostReached` and `KeyboardInterrupt` are handled by the argmin framework for all solvers, so I agree they deserve a specific variant.

In my first proposal `SolverExit(String)` catches all other cases, 
but I agree we have to distinguish `Converged` as a successful exit of the solver (for further automated handling).
So `SolverExit` would be used for unsuccessful/premature/abnormal exit, 
as such maybe it would be better named `NonConverged` or even `Failed` 

We would end up with 

```
pub enum TerminationReason {
    /// Reached maximum number of iterations
    MaxItersReached,
    /// Reached target cost function value
    TargetCostReached,
    /// Algorithm manually interrupted with Ctrl+C
    KeyboardInterrupt,
    /// Converged
    Converged, 
    /// Exit with given reason in non-converged state 
    Failed(String)
}
```
so basically I agree with you 😄

**Comment by relf** [comment]
Thanks for your feedback and smooth PR integrations!
